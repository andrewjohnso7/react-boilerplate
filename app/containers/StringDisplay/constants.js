export const FETCH_STRINGS = 'FETCH_STRINGS';
export const API_SUCCESS = 'API_SUCCESS';
export const API_FAILURE = 'API_FAILURE';
