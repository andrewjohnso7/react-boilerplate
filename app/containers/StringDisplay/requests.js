export const fetchStrings = () =>
  fetch(`api/getStrings`)
    .then(res => res.json())
    .catch(error => `Oops ${error}`);

export const updateStrings = post =>
  fetch(`/api/addString`, {
    method: 'POST',
    headers: {
      'content-type': 'application/json',
    },
    body: JSON.stringify(post),
  })
    .then(res => res.json())
    .catch(error => error);
