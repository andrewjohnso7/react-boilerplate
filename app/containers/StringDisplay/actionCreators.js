import { FETCH_STRINGS, UPDATING_STRINGS } from './constants';
/* Defining mapDispatchToProps as An Object: 
  https://react-redux.js.org/6.x/using-react-redux/connect-mapdispatch#defining-mapdispatchtoprops-as-an-object
*/

// This action is designated to fetch all strings located on the server
export const fetchStrings = () => ({ type: FETCH_STRINGS });

// This action is designated to update the strings located on the server, with the payload given
export const addString = payload => ({
  type: UPDATING_STRINGS,
  payload,
});
