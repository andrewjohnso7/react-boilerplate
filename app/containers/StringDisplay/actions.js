import { FETCH_STRINGS } from './constants';
export const fetchStrings = () => dispatch => {
  fetch(`api/getStrings`)
    .then(res => res.json())
    .then(strings =>
      dispatch({
        type: FETCH_STRINGS,
        payload: strings,
      }),
    )
    .then(console.log('We made it'))
    .catch(error => console.log('Something Broke ', error));
};
