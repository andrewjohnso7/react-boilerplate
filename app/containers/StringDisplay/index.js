import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { getSelectorStrings } from './selectors';
import { FETCH_STRINGS } from './constants';

const getStrings = () => ({ type: FETCH_STRINGS });

class StringDisplay extends React.Component {
  componentDidMount() {
    this.props.getStrings();
  }

  render() {
    return (
      <div>
        Hello
        <div>
          These are the strings
          <ul>
            {this.props.strings.strings.map(item => (
              <li key={item.id}>{item.body}</li>
            ))}
          </ul>
          <button type="button">
            <Link to="/addString">Add another string</Link>
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  strings: getSelectorStrings(state),
});

const mapDispatchToProps = {
  getStrings,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(StringDisplay);
