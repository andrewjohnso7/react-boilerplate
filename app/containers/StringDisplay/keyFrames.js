import { keyframes } from 'styled-components';
import { COLORS } from './constants';

export const backgroundColorChange = keyframes`
  from {
    background:${COLORS.mainBackground}
  }
  to {
    background:${COLORS.title}
  }
`;
