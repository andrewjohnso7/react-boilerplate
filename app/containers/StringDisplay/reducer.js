import { FETCH_STRINGS, FETCH_SUCCESS } from './constants';

const initialState = {
  list: [
    { id: 5, body: 'Aang' },
    { id: 4, body: 'Katara' },
    { id: 3, body: 'Toph' },
    { id: 2, body: 'Sokka' },
    { id: 1, body: 'Appa' },
  ],
};

export default function(state = initialState, action) {
  switch (action.type) {
    case FETCH_STRINGS:
      return {
        ...state,
      };
    case FETCH_SUCCESS: {
      return {
        ...state,
        list: action.list,
      };
    }
    default:
      return state;
  }
}
