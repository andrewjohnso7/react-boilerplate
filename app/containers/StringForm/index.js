import React from 'react';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import { addString } from '../StringDisplay/actionCreators';
import {
  MainContainer,
  Section,
  StringNavigation,
  SubmitString,
  FormContainer,
  TextArea,
} from '../StringDisplay/styled';

export class StringForm extends React.Component {
  static propTypes = {
    addString: PropTypes.func.isRequired,
    added: PropTypes.shape({
      id: PropTypes.number,
      body: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    }),
  };

  constructor(props) {
    super(props);

    this.state = {
      body: '',
    };
  }

  onChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  onSubmit = e => {
    e.preventDefault();

    const post = {
      body: this.state.body,
    };
    if (
      post.body &&
      post.body !== 'Please enter a string' &&
      post.body.length > 0
    ) {
      this.props.addString(post);
      this.setState({ body: '' });
    } else {
      this.setState({ body: 'Please enter a string' });
    }
  };

  render() {
    const { added } = this.props;
    return (
      <MainContainer>
        <Section>
          <h1>Add a Name</h1>
          <form onSubmit={this.onSubmit}>
            <FormContainer>
              <label htmlFor="New String">
                <br />
                {added ? (
                  <div>{added.body} successfully added to the list</div>
                ) : (
                  <div>Please attempt to enter a string</div>
                )}
                <br />{' '}
              </label>
              <br />
              <TextArea
                placeholder={this.state.body || "What's your name?"}
                name="body"
                value={this.state.body}
                onChange={this.onChange}
              />
              <br />
              <SubmitString type="submit">Add name</SubmitString>
            </FormContainer>
          </form>
          <br />
          <StringNavigation to="/">See list of names</StringNavigation>
        </Section>
      </MainContainer>
    );
  }
}

const mapStateToProps = state => ({
  added: state.strings.added,
  error: state.strings.error,
});

export const mapDispatchToProps = {
  addString,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(StringForm);
