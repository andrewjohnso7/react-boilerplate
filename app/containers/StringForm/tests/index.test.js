import { render } from 'react-testing-library';
import React from 'react';
import StringForm, { mapDispatchToProps } from '../index';
import { UPDATING_STRINGS } from '../../StringDisplay/constants';
/**
 *  There is only one reducer and constants file because the only state change
 *  that needs to render currently is on the StringDisplay component
 *  */

describe('StringForm component', () => {
  describe('<StringForm />', () => {
    it('should exist', () => {
      expect(StringForm).toBeDefined();
    });
    it('renders and matches the snapshot', () => {
      const { container } = render(<StringForm />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('UPDATING_STRINGS', () => {
    it('should return the correct constant', () => {
      expect(UPDATING_STRINGS).toEqual('UPDATING_STRINGS');
    });
  });

  describe('mapDispatchToProps', () => {
    it('should return the action creator for UPDATING_STRINGS', () => {
      const spy = jest.spyOn(mapDispatchToProps, 'addString');
      expect(spy).toBeDefined();
    });
  });
});
