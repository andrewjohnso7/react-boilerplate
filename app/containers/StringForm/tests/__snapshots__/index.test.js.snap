// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StringForm component <StringForm /> renders and matches the snapshot 1`] = `
<main
  class="styled__MainContainer-sc-1lglkjr-0 kEBlae"
>
  <section
    class="styled__Section-sc-1lglkjr-1 mSDmX"
  >
    <h1>
      Add a Name
    </h1>
    <form>
      <div
        class="styled__FormContainer-sc-1lglkjr-4 cXQrbI"
      >
        <label
          for="New String"
        >
          <br />
          <div>
            Please attempt to enter a string
          </div>
          <br />
           
        </label>
        <br />
        <textarea
          class="styled__TextArea-sc-1lglkjr-7 dYlekZ"
          name="body"
          placeholder="What's your name?"
        />
        <br />
        <button
          class="styled__SubmitString-sc-1lglkjr-6 dHdrXc"
          type="submit"
        >
          Add name
        </button>
      </div>
    </form>
    <br />
    <a
      class="styled__StringNavigation-sc-1lglkjr-5 kZlXKg"
      href="/"
    >
      See list of names
    </a>
  </section>
</main>
`;
